x-pack/plugin/ml/qa/native-multi-node-tests/src/test/java/org/elasticsearch/xpack/ml/integration/ResultsIndexUpgradeIT.java

```diff
@@ -0,0 +1,380 @@
+        client().admin().indices().prepareCreate(alreadyMigratedWriteIndex).execute().actionGet();
+
+        IndexNameExpressionResolver indexNameExpressionResolver = new IndexNameExpressionResolver();
+
+        ResultsIndexUpgradeService resultsIndexUpgradeService = new ResultsIndexUpgradeService(indexNameExpressionResolver,
+            logger,
+            ThreadPool.Names.SAME,
+            //indicates that this manually created index is already migrated and should not be included in our migration steps
+            indexMetaData -> !(indexMetaData.getIndex().getName().equals(alreadyMigratedIndex) ||
+                indexMetaData.getIndex().getName().equals(alreadyMigratedWriteIndex)));
+
+        PlainActionFuture<AcknowledgedResponse> future = PlainActionFuture.newFuture();
+
+        resultsIndexUpgradeService.upgrade(ESIntegTestCase.client(),
+            new ResultsIndexUpgradeAction.Request(),
+            ESIntegTestCase.client().admin().cluster().prepareState().get().getState(),
+            future);
+
+        AcknowledgedResponse response = future.get();
+        assertThat(response.isAcknowledged(), is(true));
+
+        assertThat(indexExists(job1Index), is(false));
+        assertThat(indexExists(job2Index), is(false));
+        assertThat(indexExists(job3Index), is(false));
+
+        ClusterState state = admin().cluster().state(new ClusterStateRequest()).actionGet().getState();
+        String[] indices = indexNameExpressionResolver.concreteIndexNames(state,
+            IndicesOptions.strictExpandOpenAndForbidClosed(),
+            AnomalyDetectorsIndex.jobResultsIndexPrefix() + "*");
+
+        // Our backing index size should be four as we have a shared and custom index and upgrading doubles the number of indices
+        Assert.assertThat(indices.length, equalTo(4));
+        refresh(indices);
+
+        assertThat(getJobResultsCount(job1.getId()), equalTo(job1Total));
+        assertThat(getJobResultsCount(job2.getId()), equalTo(job2Total));
+        assertThat(getJobResultsCount(job3.getId()), equalTo(job3Total));
+
+        // WE should still be able to write, and the aliases should allow to read from the appropriate indices
+        postDataToJob(jobId1);
+        postDataToJob(jobId2);
+        postDataToJob(jobId3);
+
+        refresh(indices);
+
+        long newJob1Total = getJobResultsCount(job1.getId());
+        assertThat(newJob1Total, greaterThan(job1Total));
+
+        long newJob2Total = getJobResultsCount(job2.getId());
+        assertThat(newJob2Total, greaterThan(job2Total));
+
+        long newJob3Total = getJobResultsCount(job3.getId());
+        assertThat(newJob3Total, greaterThan(job3Total));
+    }
+
+    private long getTotalDocCount(String indexName) {
+        SearchResponse searchResponse = ESIntegTestCase.client().prepareSearch(indexName)
+            .setSize(10_000)
+            .setQuery(QueryBuilders.matchAllQuery())
```

Comment:
Might be best to add `.setTrackTotalHits(true)` so this doesn't break when the default is changed to `false`.